modules/compiler/builtins/abacus/source/abacus_math/ldexp.cpp

```diff
@@ -20,312 +20,119 @@
-  T result = __abacus_select(x * mul_hi, x_fract * mul_hi, cond_frexp);
-  // Parens since floating point operations aren't associative, and any other
-  // order would lose precision from x
-  result = (result * mul_lo) * mul_lo_lo;
-
-  // check if n is very big or very small:
-  // The bound is the value of n necessary to go from the smallest number to the
-  // biggest one:
-  //   - 2 * 15: n to go from the smallest normal number to the biggest.
-  //   - 9: n to go from the smallest denormal back into the normal range.
-  //   (smallest denormal is 2 ^ -24, so 24 - 15 = 9).
-  //   bound ==> (2 * 15) + 9 ==> 39
-  const S normalize_n = 9;
-  const S bias_as_int = abacus::detail::cast::convert<S>(exp_bias);
-  const S bound = 2 * bias_as_int + normalize_n;
-
-  // Underflow is defined as returning signed zero
-  const SignedType cond2 =
-      abacus::detail::cast::convert<SignedType>(n < -bound);
-  const T copy_sign = __abacus_copysign(0.0f16, x);
-  result = __abacus_select(result, copy_sign, cond2);
-
-  // Overflow is defined as returning signed inf
-  const SignedType inf = Shape::ExponentMask();
-  const T half_infinity = abacus::detail::cast::as<T>(inf);
-  const T copy_sign2 = __abacus_copysign(half_infinity, x);
-  const SignedType cond3 = abacus::detail::cast::convert<SignedType>(n > bound);
-  result = __abacus_select(result, copy_sign2, cond3);
-
-  T return_value = x;
-  const SignedType isnan = __abacus_isnan(x);
-
-  // If flush to zero is supported and this is a denormal number then we should
-  // flush to zero.
-  const SignedType is_denorm = abacus::internal::is_denorm(x);
-  if (__abacus_isftz()) {
-    return_value = __abacus_select(return_value, (T)0.0f16, is_denorm);
-    // Is denorm can be true for nan so ignore it in that case.
-    return_value = __abacus_select(x, return_value, isnan);
-  }
-
-  const SignedType is_zero = (x == 0.0f16) & ~is_denorm;
-  SignedType cond4 = __abacus_isinf(x) | isnan | is_zero;
-
-  return __abacus_select(result, return_value, cond4);
-}
-#endif
-
-template <typename T, typename SignedType = typename TypeTraits<T>::SignedType>
-inline T ldexp_float(const T x, const SignedType n) {
-  typedef FPShape<T> Shape;
-
-  // split n into 3 depending on how big it is
-  const abacus_int exp_bias = Shape::Bias();
-  SignedType n_hi = __abacus_select(n, n / 3, __abacus_abs(n) > (exp_bias - 1));
-  SignedType n_lo = __abacus_select(0, n_hi, __abacus_abs(n) > (exp_bias - 1));
-  SignedType n_lo_lo = n - (n_hi + n_lo);
-
-  // put these n's into exponents:
-  n_hi += exp_bias;
-  n_lo += exp_bias;
-  n_lo_lo += exp_bias;
+  const SignedType n_c = convert_sat<SignedType>(n);
 
-  const SignedType mant_bits = Shape::Mantissa();
-  T mul_hi = abacus::detail::cast::as<T>(n_hi << mant_bits);
-  T mul_lo = abacus::detail::cast::as<T>(n_lo << mant_bits);
-  T mul_lo_lo = abacus::detail::cast::as<T>(n_lo_lo << mant_bits);
+  // Split n into 3 such that ldexp(ldexp(x, n_1), n_1) does not turn normal
```

Comment:
could we explain what split n into three really means and how the algorithm works, possibly with a link?

Reply:
that's a good explanation, can you add that?

Reply:
Thanks for the explanation, makes sense (I was also unsure).

If you could add some version of this comment to the source code that'd be really helpful.

Reply:
Expanded the comment, does it look okay like this?

Reply:
yes that is much better - I've added a comment below though just to add a bit more so it's obvious that ldexp -> x * pow(n)